rclcpp/src/rclcpp/sub_node.cpp

```diff
@@ -0,0 +1,321 @@
+{
+  return node_graph_->count_subscribers(topic_name);
+}
+
+const std::vector<rclcpp::callback_group::CallbackGroup::WeakPtr> &
+SubNode::get_callback_groups() const
+{
+  return node_base_->get_callback_groups();
+}
+
+rclcpp::Event::SharedPtr
+SubNode::get_graph_event()
+{
+  return node_graph_->get_graph_event();
+}
+
+void
+SubNode::wait_for_graph_change(
+  rclcpp::Event::SharedPtr event,
+  std::chrono::nanoseconds timeout)
+{
+  node_graph_->wait_for_graph_change(event, timeout);
+}
+
+rclcpp::Clock::SharedPtr
+SubNode::get_clock()
+{
+  return node_clock_->get_clock();
+}
+
+rclcpp::Time
+SubNode::now()
+{
+  return node_clock_->get_clock()->now();
+}
+
+rclcpp::node_interfaces::NodeBaseInterface::SharedPtr
+SubNode::get_node_base_interface()
+{
+  return node_base_;
+}
+
+rclcpp::node_interfaces::NodeClockInterface::SharedPtr
+SubNode::get_node_clock_interface()
+{
+  return node_clock_;
+}
+
+rclcpp::node_interfaces::NodeGraphInterface::SharedPtr
+SubNode::get_node_graph_interface()
+{
+  return node_graph_;
+}
+
+rclcpp::node_interfaces::NodeTimersInterface::SharedPtr
+SubNode::get_node_timers_interface()
+{
+  return node_timers_;
+}
+
+rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr
+SubNode::get_node_topics_interface()
+{
+  return node_topics_;
+}
+
+rclcpp::node_interfaces::NodeServicesInterface::SharedPtr
+SubNode::get_node_services_interface()
+{
+  return node_services_;
+}
+
+rclcpp::node_interfaces::NodeParametersInterface::SharedPtr
+SubNode::get_node_parameters_interface()
+{
+  return node_parameters_;
+}
+
+const std::string
+SubNode::validated_extended_ns(const std::string & extended_ns) const
+{
+  std::string ret_ns = extended_ns;
+
+  if (ret_ns[0] == '/') {
+    ret_ns.erase(ret_ns.begin());
+  }
+  // converts "//" to "/"
+  std::string::size_type found = ret_ns.find("//");
+  while (found != std::string::npos) {
+    ret_ns.replace(found, 2, "/");
+    found = ret_ns.find("//");
+  }
+
+  std::string full_namespace = std::string(original_->get_namespace()) + "/" + ret_ns;
+
+  int validation_result;
+  size_t invalid_index;
+  rmw_ret_t rmw_ret =
+    rmw_validate_namespace(full_namespace.c_str(), &validation_result, &invalid_index);
+
+  if (rmw_ret != RMW_RET_OK) {
+    if (rmw_ret == RMW_RET_INVALID_ARGUMENT) {
+      throw_from_rcl_error(RCL_RET_INVALID_ARGUMENT, "failed to validate subnode namespace");
+    }
+    throw_from_rcl_error(RCL_RET_ERROR, "failed to validate subnode namespace");
+  }
+
+  if (validation_result != RMW_NAMESPACE_VALID) {
+    throw rclcpp::exceptions::InvalidNamespaceError(
+            full_namespace.c_str(),
+            rmw_namespace_validation_result_string(validation_result),
+            invalid_index);
+  }
+
+  return ret_ns;
+}
+
+RCLCPP_PUBLIC
+const std::string
+SubNode::add_extended_ns_prefix(const std::string & original_name) const
+{
+  return std::string(original_->get_namespace()) + "/" + extended_namespace_ + "/" + original_name;
```

Comment:
This works for relative names `foo` or `foo/bar`, but I'm not sure about the behavior for fully qualified or private names `~/bar`.

ROS 1 threw `ros::InvalidNameException` when trying to create a publisher/subscriber/etc with a private name, asking instead that the `NodeHandle` be created with a private namespace. In ROS 2 the `create_...()` APIs allow private names. I'm not sure how to handle all the cases, but there is a proposal here: https://github.com/ros2/rclcpp/issues/531#issuecomment-437201989